New feature for Salt-Cohesity Integration

naveena-maplelabs opened the pull request:
Integrate cohesity with salt.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltstack.com/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [ ] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes/No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

naveena-maplelabs committed:
Python code enhancement

naveena-maplelabs committed:
Python Salt integration module

naveena-maplelabs committed:
Code Enhancement:Updated python module

naveena-maplelabs committed:
Salt Enhancement

naveena-maplelabs committed:
Module enhancement

naveena-maplelabs committed:
Integration Module enhancement

naveena-maplelabs committed:
Cohesity Integration: module enhancement

naveena-maplelabs committed:
Added doc for cohesity module

naveena-maplelabs committed:
Updated index.rst file

naveena-maplelabs committed:
Add requirements

naveena-maplelabs committed:
Add windows requirements

naveena-maplelabs committed:
Fixed import errors

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs committed:
Added cohesity.rst file

naveena-maplelabs committed:
Updated index.rst file

naveena-maplelabs committed:
Updated rst file

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs committed:
Add __virtual__ function

naveena-maplelabs committed:
Pylint fix

naveena-maplelabs committed:
Pylint fix

naveena-maplelabs committed:
Updated requirement files

naveena-maplelabs committed:
Removed json pickle package version

naveena-maplelabs committed:
Added lint fix

naveena-maplelabs committed:
Added lint fix

naveena-maplelabs commented:
@s0undt3ch, can you please point out what could be the reason for these pre-commit errors and how can i fix the same.

naveena-maplelabs committed:
Updated module

s0undt3ch commented:
```shell
pip install pre-commit
pre-commit run --files <the> <files> <you> <changed>
```

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs committed:
Updated requirements file

naveena-maplelabs committed:
Requirements file updated for py3.5

naveena-maplelabs committed:
Updated requirements files

naveena-maplelabs committed:
Requirement error fix

naveena-maplelabs committed:
Resolve requirement files error

naveena-maplelabs committed:
Resolve requirement files error

naveena-maplelabs committed:
Add deleted requirement files

naveena-maplelabs committed:
Requirement files updated

naveena-maplelabs committed:
Added removed requirements txt files

Commit:
Requiremnet packages updated

naveena-maplelabs committed:
Requirement package added

naveena-maplelabs committed:
Requirement package added

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs committed:
Removed unwanted txt files

naveena-maplelabs committed:
Requirements packages added

naveena-maplelabs committed:
Updated requirements

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs committed:
Remove jsonpickle package version conflcit

naveena-maplelabs committed:
Updated requirements

Commit:
Removed .txt files

naveena-maplelabs committed:
Updated requirements

naveena-maplelabs committed:
Updated requirements

naveena-maplelabs committed:
Updated requirements

naveena-maplelabs committed:
Updated requirements

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs committed:
Updated requirements

naveena-maplelabs committed:
Updated requirements

naveena-maplelabs committed:
Updated requirements

naveena-maplelabs committed:
Updated requirements

naveena-maplelabs commented:
My new module for cohesity-salt integration is not loaded correctly. i m getting following error. I have added the module under salt/salt/modules folder. @s0undt3ch Do we need to updated any other file?

Warning, treated as error:
[autosummary] failed to import 'salt.modules.cohesity': no module named salt.modules.cohesity
Failed to import 'salt.modules.cohesity': no module named salt.modules.cohesity
make: *** [Makefile:56: html] Error 2

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs committed:
Updated requirements

naveena-maplelabs committed:
Updated requirements

naveena-maplelabs committed:
Updated requirements

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs committed:
Updated requirements

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs committed:
Updated requirements

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs committed:
Updated requirements

naveena-maplelabs committed:
Updated requirements

naveena-maplelabs committed:
Updated requirements

naveena-maplelabs committed:
Updated requirements

naveena-maplelabs committed:
Updated requirements

naveena-maplelabs committed:
Updated requirements

naveena-maplelabs committed:
Updated requirements

naveena-maplelabs committed:
Updated requirements

naveena-maplelabs committed:
Updated requirements

naveena-maplelabs committed:
Updated requirements

naveena-maplelabs committed:
Updated requirements

naveena-maplelabs committed:
Updated module - lint workflow run

naveena-maplelabs committed:
Updated module - lint workflow run

naveena-maplelabs committed:
Updated module - lint workflow run

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs committed:
Updated module - lint workflow run

naveena-maplelabs committed:
Updated module - lint workflow run

naveena-maplelabs committed:
Updated module - lint workflow run

naveena-maplelabs committed:
Updated module - lint workflow run

naveena-maplelabs committed:
Updated module - lint workflow run

naveena-maplelabs committed:
Updated module - lint workflow run

naveena-maplelabs committed:
Updated module - lint workflow run

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs committed:
Updated Cohesity docstring

naveena-maplelabs committed:
Updated Cohesity docstring

naveena-maplelabs committed:
Updated Cohesity docstring

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs committed:
Updated SDK version

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs commented:
@s0undt3ch All the checks are passed, Can you please review the code.

s0undt3ch commented on doc/man/salt.7:
```diff
@@ -110489,6 +110489,12 @@ A module for shelling out.
 T}
 _
 T{
+\fBcohesity\fP
```
Don't change manpages. That's done by us right before a release

naveena-maplelabs replied:
Okay

s0undt3ch commented on requirements/base.txt:
```diff
@@ -4,4 +4,4 @@ PyYAML
 requests>=1.0.0
 distro>=1.5
-contextvars; python_version < '3.7'
+contextvars; python_version < '3.7'
```
What was this change? Github's diff doesn't help much....

s0undt3ch commented on requirements/pytest.txt:
```diff
@@ -5,4 +5,4 @@ pytest-salt-factories >= 0.121.1
 pytest-helpers-namespace >= 2019.1.8
 pytest-subtests
-flaky
+flaky
```
What was this change? Github's diff doesn't help much....

s0undt3ch commented on requirements/static/ci/docs.in:
```diff
@@ -1 +1,2 @@
 sphinx>=3.0.1,!=3.0.2
+cohesity-management-sdk==1.6.0
```
Do we really need this for docs? Imports can be mocked in `docs/conf.py`

s0undt3ch commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,522 @@
+sdk can be installed using `pip install cohesity-management-sdk`
+"""
+
+# Python Modules Import
```
```suggestion
```

pyashish commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,416 @@
+c_domain = cohesity_config['domain']
+
+cohesity_client = CohesityClient(cluster_vip=cluster_vip,
+                                         username=c_username,
```
intend

naveena-maplelabs replied:
Done

pyashish commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,416 @@
+        cohesity_client.protection_sources.create_register_protection_source(
+            body)
+        return 'Successfully registered Vcenter %s' % vcenter
+    except Exception as err:
```
Can we do APIExceptions for all exceptions ?

s0undt3ch commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,522 @@
+    from cohesity_management_sdk.exceptions.api_exception import APIException
+    from cohesity_management_sdk.models.cancel_protection_job_run_param import (
+        CancelProtectionJobRunParam,
+    )
+    from cohesity_management_sdk.models.change_protection_job_state_param import (
+        ChangeProtectionJobStateParam,
+    )
+    from cohesity_management_sdk.models.delete_protection_job_param import (
+        DeleteProtectionJobParam,
+    )
+    from cohesity_management_sdk.models.environment_register_protection_source_parameters_enum import (
+        EnvironmentRegisterProtectionSourceParametersEnum as env_enum,
+    )
+    from cohesity_management_sdk.models.protection_job_request_body import (
+        ProtectionJobRequestBody,
+    )
+    from cohesity_management_sdk.models.recover_task_request import RecoverTaskRequest
+    from cohesity_management_sdk.models.register_protection_source_parameters import (
+        RegisterProtectionSourceParameters,
+    )
+    from cohesity_management_sdk.models.restore_object_details import (
+        RestoreObjectDetails,
+    )
+    from cohesity_management_sdk.models.run_protection_job_param import (
+        RunProtectionJobParam,
+    )
+    from cohesity_management_sdk.models.universal_id import UniversalId
+    from cohesity_management_sdk.models.update_protection_jobs_state_request_body import (
+        UpdateProtectionJobsStateRequestBody,
+    )
+    from cohesity_management_sdk.models.vmware_restore_parameters import (
+        VmwareRestoreParameters,
+    )
+
+    HAS_LIBS = True
+except ImportError as err:
+    HAS_LIBS = False
+    print("Error while importing Cohesity SDK modules.")
+    print(err)
+    exit(0)
```
```suggestion
import salt.utils.files
try:

    from cohesity_management_sdk.cohesity_client import CohesityClient
    from cohesity_management_sdk.exceptions.api_exception import APIException
    from cohesity_management_sdk.models.cancel_protection_job_run_param import (
        CancelProtectionJobRunParam,
    )
    from cohesity_management_sdk.models.change_protection_job_state_param import (
        ChangeProtectionJobStateParam,
    )
    from cohesity_management_sdk.models.delete_protection_job_param import (
        DeleteProtectionJobParam,
    )
    from cohesity_management_sdk.models.environment_register_protection_source_parameters_enum import (
        EnvironmentRegisterProtectionSourceParametersEnum as env_enum,
    )
    from cohesity_management_sdk.models.protection_job_request_body import (
        ProtectionJobRequestBody,
    )
    from cohesity_management_sdk.models.recover_task_request import RecoverTaskRequest
    from cohesity_management_sdk.models.register_protection_source_parameters import (
        RegisterProtectionSourceParameters,
    )
    from cohesity_management_sdk.models.restore_object_details import (
        RestoreObjectDetails,
    )
    from cohesity_management_sdk.models.run_protection_job_param import (
        RunProtectionJobParam,
    )
    from cohesity_management_sdk.models.universal_id import UniversalId
    from cohesity_management_sdk.models.update_protection_jobs_state_request_body import (
        UpdateProtectionJobsStateRequestBody,
    )
    from cohesity_management_sdk.models.vmware_restore_parameters import (
        VmwareRestoreParameters,
    )

    HAS_LIBS = True
except ImportError:
    HAS_LIBS = False
```

s0undt3ch commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,522 @@
+    print(err)
+    exit(0)
+
+logger = logging.getLogger(__name__)
```
```suggestion
log = logging.getLogger(__name__)
```
Please use `log` and not `logger` since that's what's used throughout Salt's codebase.

naveena-maplelabs replied:
Done

pyashish commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,416 @@
+
+def create_vmware_protection_job(job_name, vcenter_name, sources,
+    policy='Gold', domain='DefaultStorageDomain', pause_job=True,
+    timezone='Europe/Berlin', description=''):
```
Use local time from where salt is running if not specified :  
`'/'.join(os.path.realpath('/etc/localtime').split('/')[-2:]) `

pyashish commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,416 @@
+    timezone='Europe/Berlin', description=''):
+    '''
+    Create Virtual Protection Source
+    :param jobname
```
Should have more description because this is going into docs

s0undt3ch commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,522 @@
+ERROR_LIST = []
+__virtualname__ = "cohesity"
+
+config_path = "/etc/salt/master.d/cohesity.conf"
```
This just looks wrong. It will ignore when `--config-dir` is passed from CLIs?

pyashish commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,416 @@
+        body.objects = []
+        source_objects = \
+            cohesity_client.protection_sources.list_protection_sources(
+            id=parent_id, include_datastores=True, exclude_types=['kHostSystem',
```
intend, follow pep8

s0undt3ch commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,522 @@
+__virtualname__ = "cohesity"
+
+config_path = "/etc/salt/master.d/cohesity.conf"
+cohesity_config = {}
+if os.path.isfile(config_path):
+    import yaml
+
+    with salt.utils.files.fopen(config_path, "r") as file_obj:
+        config = yaml.safe_load(file_obj)
+        cohesity_config = config.get("cohesity_config", {})
+
+cluster_vip = cohesity_config.get("cluster_vip", "")
+c_username = cohesity_config.get("username", "")
+c_password = cohesity_config.get("password", "")
+c_domain = cohesity_config.get("domain", "")
+cohesity_client = CohesityClient(
+    cluster_vip=cluster_vip, username=c_username, password=c_password, domain=c_domain,
+)
```
This should be deferred to a helper function, which then has access to `__opts__` so you don't have to look for a hardcoded path to the config.

s0undt3ch replied:
```python
def _get_client():
    context_key = '{}.cohesity_client'.format(__name__)
    if context_key in __context__:
        return __context__[context_key]
    cohesity_config = config.get("cohesity_config", {})
    cluster_vip = cohesity_config.get("cluster_vip", "")
    c_username = cohesity_config.get("username", "")
    c_password = cohesity_config.get("password", "")
    c_domain = cohesity_config.get("domain", "")
    cohesity_client = CohesityClient(
        cluster_vip=cluster_vip, username=c_username, password=c_password, domain=c_domain,
    )
    __context__[context_key] = cohesity_client
    return cohesity_client
```

naveena-maplelabs replied:
Updated

pyashish commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,416 @@
+        source_objects = \
+            cohesity_client.protection_sources.list_protection_sources(
+            id=parent_id, include_datastores=True, exclude_types=['kHostSystem',
+            'kVirtualMachine'], environment=env_enum.K_VMWARE)
```
Some places enums models are used vs static. Make sure its consitent

naveena-maplelabs committed:
Minor Fix

naveena-maplelabs committed:
Minor Fix

naveena-maplelabs committed:
Minor Fix

naveena-maplelabs committed:
Updated requirement

naveena-maplelabs committed:
Updated requirement

naveena-maplelabs commented:
@s0undt3ch I need to add a dependency python-pip package 'cohesity-management-sdk'. where should i add this requirement? Previously i have added in requirements/static/ci/docs.in file. After your comment i have removed from docs.in file. But the package is not installed now.

naveena-maplelabs committed:
Updated files

s0undt3ch commented:
@naveena-maplelabs you need it for testing purposes, correct?
If so, and if it's supposed to be installed for all platforms, the file is `requirements/static/ci/common.in`, or if it should only be installed for a specific platform, `requirements/static/ci/<platform-name>.in`(This file must already exist)

naveena-maplelabs committed:
Updated files

naveena-maplelabs committed:
Updated files

naveena-maplelabs committed:
Updated files

naveena-maplelabs committed:
Updated files

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs committed:
Updated files

naveena-maplelabs committed:
Updated files

naveena-maplelabs committed:
Updated files

naveena-maplelabs committed:
Updated files

naveena-maplelabs committed:
Updated files

naveena-maplelabs committed:
Updated files

naveena-maplelabs committed:
Updated files

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs commented:
> @naveena-maplelabs you need it for testing purposes, correct?
> If so, and if it's supposed to be installed for all platforms, the file is `requirements/static/ci/common.in`, or if it should only be installed for a specific platform, `requirements/static/ci/<platform-name>.in`(This file must already exist)

Okay, thanks

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs commented:
@s0undt3ch I have addressed your comments and updated the script. Can you please review it.

naveena-maplelabs commented:
@s0undt3ch can you please review the code.

s0undt3ch commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,554 @@
+    HAS_LIBS = True
+except ImportError as err:
+    HAS_LIBS = False
+    print("Error while importing Cohesity SDK modules.")
+    print(err)
```
```suggestion
```

s0undt3ch commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,554 @@
+# )
+
+
+def __virtual__():
+    if HAS_LIBS:
+        return __virtualname__
+    return True
```
```suggestion
def __virtual__():
    if HAS_LIBS:
        return __virtualname__
    return False, "Could not import the Cohesity SDK modules"
```

s0undt3ch commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,554 @@
+    )
+    __context__[context_key] = cohesity_client
+    return cohesity_client
+
+
+# config_path = "/etc/salt/master.d/cohesity.conf"
+# cohesity_config = {}
+# if os.path.isfile(config_path):
+#     import yaml
+#
+#     with salt.utils.files.fopen(config_path, "r") as file_obj:
+#         config = yaml.safe_load(file_obj)
+#         cohesity_config = config.get("cohesity_config", {})
+#
+# cluster_vip = cohesity_config.get("cluster_vip", "")
+# c_username = cohesity_config.get("username", "")
+# c_password = cohesity_config.get("password", "")
+# c_domain = cohesity_config.get("domain", "")
+# cohesity_client = CohesityClient(
+#     cluster_vip=cluster_vip, username=c_username, password=c_password, domain=c_domain,
+# )
```
```suggestion
```

s0undt3ch commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,554 @@
+
+
+def get_sd_id(name):
+    """
+    Function to fetch storage domain available in the cluster.
+    : return storage domain id.
+    """
```
```suggestion
    """
    Function to fetch storage domain available in the cluster.
    
    :return: storage domain id.
    """
```

s0undt3ch replied:
Apply the same to all of you docstrings which have `: return (... etc ...)`

s0undt3ch commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,554 @@
+        return -1, []
+
+
+def register_vcenter(vcenter, username, password):
+    """
+    Function to fetch register Vmware Vcenter to cohesity cluster.
+
+    CLI Examples:
+
+    .. code-block:: bash
+
+    salt-call cohesity.register_vcenter vcenter=vcenter_name username=admin password=admin
+    """
```
```suggestion
def register_vcenter(vcenter, username, password):
    """
    Function to fetch register Vmware Vcenter to cohesity cluster.

    CLI Examples:

    .. code-block:: bash

        salt-call cohesity.register_vcenter vcenter=vcenter_name username=admin password=admin
    """
```

s0undt3ch replied:
Apply the same ident on all your other `.. code-block:: bash`

naveena-maplelabs replied:
Done

naveena-maplelabs committed:
Update salt/modules/cohesity.py

Co-authored-by: Pedro Algarvio <pedro@algarvio.me>

naveena-maplelabs committed:
Update salt/modules/cohesity.py

Co-authored-by: Pedro Algarvio <pedro@algarvio.me>

naveena-maplelabs committed:
Update salt/modules/cohesity.py

Co-authored-by: Pedro Algarvio <pedro@algarvio.me>

naveena-maplelabs committed:
Update salt/modules/cohesity.py

Co-authored-by: Pedro Algarvio <pedro@algarvio.me>

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs committed:
Updated files

naveena-maplelabs commented:
@s0undt3ch thanks for reviewing. I have addressed your comments.

naveena-maplelabs committed:
Updated files

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs commented:
@s0undt3ch Addressed your comments. Please review the code. Thanks

s0undt3ch commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,537 @@
+    HAS_LIBS = False
+
+log = logging.getLogger(__name__)
+ERROR_LIST = []
```
```suggestion
```
You don't seem to use this.

s0undt3ch commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,537 @@
+log = logging.getLogger(__name__)
+ERROR_LIST = []
+__virtualname__ = "cohesity"
+__context__ = {}
```
```suggestion
```
Salt will inject `__context__` when the module loads.

s0undt3ch commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,537 @@
+__context__ = {}
+
+
+def _get_client():
+    context_key = "{}.cohesity_client".format(__name__)
+    opts = salt.config.master_config("/etc/salt/master")
+    conf = opts.get("conf_file", "")
+    if context_key in __context__:
+        return __context__[context_key]
+    cohesity_config = opts.get("cohesity_config", {})
+    cluster_vip = cohesity_config.get("cluster_vip", "")
+    c_username = cohesity_config.get("username", "")
+    c_password = cohesity_config.get("password", "")
+    c_domain = cohesity_config.get("domain", "")
+    cohesity_client = CohesityClient(
+        cluster_vip=cluster_vip,
+        username=c_username,
+        password=c_password,
+        domain=c_domain,
+    )
+    __context__[context_key] = cohesity_client
+    return cohesity_client
```
```suggestion
def _get_client():
    context_key = "{}.cohesity_client".format(__name__)
    if context_key in __context__:
        return __context__[context_key]
    cohesity_config = __opts__.get("cohesity_config", {})
    cluster_vip = cohesity_config.get("cluster_vip", "")
    c_username = cohesity_config.get("username", "")
    c_password = cohesity_config.get("password", "")
    c_domain = cohesity_config.get("domain", "")
    cohesity_client = CohesityClient(
        cluster_vip=cluster_vip,
        username=c_username,
        password=c_password,
        domain=c_domain,
    )
    __context__[context_key] = cohesity_client
    return cohesity_client
```

naveena-maplelabs committed:
Update salt/modules/cohesity.py

Co-authored-by: Pedro Algarvio <pedro@algarvio.me>

naveena-maplelabs committed:
Update salt/modules/cohesity.py

Co-authored-by: Pedro Algarvio <pedro@algarvio.me>

naveena-maplelabs committed:
Update salt/modules/cohesity.py

Co-authored-by: Pedro Algarvio <pedro@algarvio.me>

naveena-maplelabs commented:
Okay @s0undt3ch, done.

naveena-maplelabs committed:
Updated files

naveena-maplelabs commented:
Updated the code. Please review @s0undt3ch .

s0undt3ch commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,531 @@
+    : return: storage domain id.
+    """
+    cohesity_client = _get_client()
+    log.info("Getting sorage domain with name {}".format(name))
```
```suggestion
    log.info("Getting sorage domain with name %s", name)
```

s0undt3ch commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,531 @@
+    : return policy id.
+    """
+    cohesity_client = _get_client()
+    log.info("Getting policy with name {}".format(name))
```
```suggestion
    log.info("Getting policy with name %s", name)
```

s0undt3ch commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,531 @@
+            if name in [endpoint, v_name]:
+                parent_id = each_source.protection_source.id
+        if parent_id == -1:
+            log.error("Vcenter {} not available in the cluster".format(name))
+            exit()
```
```suggestion
            log.error("Vcenter %s not available in the cluster", name)
```
Never call `exit()`, return an error or something else you need to, but never `exit`

s0undt3ch commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,531 @@
+            vm_names.remove(vm.name)
+            source_id_list.append(vm.id)
+        if vm_names:
+            log.error(
+                "Following list of vms '{}' are not available in vcenter, "
+                "please make sure the virtual machine names are correct".format(
+                    ",".join(vm_names)
+                )
+            )
```
```suggestion
            log.error(
                "Following list of vms '%s' are not available in vcenter, "
                "please make sure the virtual machine names are correct",
                ",".join(vm_names)
            )
```

s0undt3ch commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,531 @@
+        )
+
+
+def fetch_source_objects(source_objects, source_type, name=None):
```
If this is not expected to be called from the CLI, prefix it with `_`

naveena-maplelabs committed:
Updated files

naveena-maplelabs committed:
Updated files

naveena-maplelabs committed:
Updated files

naveena-maplelabs committed:
Updated files

naveena-maplelabs committed:
Updated files

naveena-maplelabs committed:
Updated files

naveena-maplelabs commented:
@s0undt3ch got lint error while replacing the string.format(), is it mandatory to replace .format() with %s.

naveena-maplelabs committed:
Merge branch 'master' into master

s0undt3ch commented:
@naveena-maplelabs regular strings should use `.format()`, log calls should use string replacement`%s`...

naveena-maplelabs committed:
Updated files

naveena-maplelabs committed:
Updated files

naveena-maplelabs commented:
@s0undt3ch Updated the log calls with %s. Please review.

naveena-maplelabs committed:
Updated files

naveena-maplelabs commented:
@s0undt3ch can you please review the code.

naveena-maplelabs committed:
Added option for user-agent

naveena-maplelabs commented:
@s0undt3ch 
`salt/modules/cohesity.py:92: [W1201(logging-not-lazy), _get_sd_id] Specify string format arguments as logging function parameters
salt/modules/cohesity.py:92: [E1321(string-substitution-usage-error), _get_sd_id] String substitution used instead of string formattting on: 'Getting storage domain with name %s'
salt/modules/cohesity.py:105: [W1201(logging-not-lazy), _get_policy_id] Specify string format arguments as logging function parameters
salt/modules/cohesity.py:105: [E1321(string-substitution-usage-error), _get_policy_id] String substitution used instead of string formattting on: 'Getting policy with name %s'
salt/modules/cohesity.py:133: [W1201(logging-not-lazy), _get_vmware_source_ids] Specify string format arguments as logging function parameters
salt/modules/cohesity.py:133: [E1321(string-substitution-usage-error), _get_vmware_source_ids] String substitution used instead of string formattting on: 'Vcenter %s not available in the cluster'
salt/modules/cohesity.py:142: [W1201(logging-not-lazy), _get_vmware_source_ids] Specify string format arguments as logging function parameters
salt/modules/cohesity.py:143: [E1321(string-substitution-usage-error), _get_vmware_source_ids] String substitution used instead of string formattting on: "Following list of vms '%s' are not available in vcenter, please make sure the virtual machine names are correct" `
Can we ignore these lint errors.

s0undt3ch commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,532 @@
+    : return: storage domain id.
+    """
+    cohesity_client = _get_client()
+    log.info("Getting storage domain with name %s" % name)
```
```suggestion
    log.info("Getting storage domain with name %s", name)
```
Sorry, this is how log calls should be done.

naveena-maplelabs replied:
Okay Thank you,

naveena-maplelabs committed:
Added lint fix

s0undt3ch commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,532 @@
+            vm_names.remove(vm.name)
+            source_id_list.append(vm.id)
+        if vm_names:
+            log.error(
+                "Following list of vms '%s' are not available in vcenter, "
+                "please make sure the virtual machine names are correct"
+               , (",".join(vm_names))
+            )
```
```suggestion
            log.error(
                "Following list of vms '%s' are not available in vcenter, "
                "please make sure the virtual machine names are correct",
                ",".join(vm_names)
            )
```

naveena-maplelabs replied:
Done

naveena-maplelabs committed:
Updated files

naveena-maplelabs committed:
Updated files

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs commented:
Thanks @s0undt3ch.

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs commented:
@s0undt3ch can you please review the code.

s0undt3ch commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,532 @@
+
+    # Update the user-agent.
+    BaseController.global_headers["user-agent"] = "cohesity-salt/1.0.0"
+except ImportError as err:
+    HAS_LIBS = False
```
```suggestion
except ImportError:
    HAS_LIBS = False
```
You're not using `err` anywhere.

naveena-maplelabs replied:
removed

s0undt3ch commented on salt/modules/cohesity.py:
```diff
@@ -0,0 +1,532 @@
+            vm_names.remove(vm.name)
+            source_id_list.append(vm.id)
+        if vm_names:
+            log.error(
+                "Following list of vms '%s' are not available in vcenter, "
+                "please make sure the virtual machine names are correct",
+                (",".join(vm_names)),
+            )
```
```suggestion
            log.error(
                "Following list of vms '%s' are not available in vcenter, "
                "please make sure the virtual machine names are correct",
                ",".join(vm_names),
            )
```

s0undt3ch reviewed (changes requested):
Some final changes.

And the PR is also lacking tests for the changes it introduces, preferable done using pytest under `tests/pytests`

naveena-maplelabs committed:
Update salt/modules/cohesity.py

Co-authored-by: Pedro Algarvio <pedro@algarvio.me>

naveena-maplelabs committed:
Updated files

naveena-maplelabs commented:
@s0undt3ch do we need to add tests under pytests/unit/modules folder.

s0undt3ch commented:
`tests/pytests/unit/modules`, yes, please.

naveena-maplelabs committed:
Added pytests for cohesity

naveena-maplelabs commented:
@s0undt3ch do we require test functions for all available methods in module, since all our methods depends on login.

s0undt3ch commented:
Preferably, yes, unless there's a compelling reason not to

pyashish commented:
Hi Pedro,

These are infra behind a VPC in a datacenter, and it'd be hard to write
tests for these. We did look at other modules, they are probably using
MagicMock. Can we make a release for now and add the pytests later using
MagicMock?  We will make sure we do it internally before making any push to
master/release for our mutual customers.
Can we get Approval for the merge please?



On Wed, Mar 17, 2021 at 11:08 PM Pedro Algarvio ***@***.***>
wrote:

> Preferably, yes, unless there's a compelling reason not to
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/saltstack/salt/pull/59328#issuecomment-801657040>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALCPI5ZGM52XGR5765VOY4LTEGKEBANCNFSM4WKD5GOA>
> .
>


-- 
Regards,
Ashish Bhat

naveena-maplelabs commented:
@s0undt3ch
These are infra behind a VPC in a datacenter, and it'd be hard to write
tests for these. We did look at other modules, they are probably using
MagicMock. Can we make a release for now and add the pytests later using
MagicMock?  We will make sure we do it internally before making any push to
master/release for our mutual customers.
Can we get Approval for the merge please?

s0undt3ch commented:
Sadly, no.
We no longer accept new code without proper test coverage. At the bare minimum, meaningful unit tests are required.

I can suggest that instead of including your module in salt, you create a salt extension.

We are still working through proper documentation for extensions but you can have a look at an example extension in https://github.com/saltstack/salt-ext-test-1

The realy important bits for an extension is https://github.com/saltstack/salt-ext-test-1/blob/master/setup.cfg#L47 and https://github.com/saltstack/salt-ext-test-1/blob/master/src/saltext/tidx1/loader.py

Once you have a salt extension package(which is basically a python package with Salt's entry-points defined), you can pip install it into the same python where salt is installed and it will get picked up by salt.

naveena-maplelabs commented:
Okay @s0undt3ch , thanks for your suggestion.
We will add the testcases.

naveena-maplelabs committed:
Sample testcase for cohesity

naveena-maplelabs committed:
Merge branch 'master' into master

naveena-maplelabs committed:
Sample testcase for cohesity

naveena-maplelabs committed:
Sample testcase for cohesity

naveena-maplelabs commented:
@s0undt3ch i have added sample testcase, can you please check it.

sagetherage commented:
@naveena-maplelabs I will get someone to review the test case, here, we would like to consider not adding this to the core of Salt code base and make this a salt-extension. We can do this in a myriad of ways, please let us know if you agree and if you want to host this in your own repo or in our GitHub org - both are possible.

pyashish commented:
End goal is to get listed here
https://docs.saltproject.io/en/latest/ref/modules/all/salt.modules.infoblox.html#module-salt.modules.infoblox
or equivalent of the extension. Please let us know what is the best
possible way.

On Tue, Apr 13, 2021 at 12:19 PM Sage the Rage ***@***.***>
wrote:

> @naveena-maplelabs <https://github.com/naveena-maplelabs> I will get
> someone to review the test case, here, we would like to consider not adding
> this to the core of Salt code base and make this a salt-extension. We can
> do this in a myriad of ways, please let us know if you agree and if you
> want to host this in your own repo or in our GitHub org - both are possible.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/saltstack/salt/pull/59328#issuecomment-818991812>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALCPI54O65UCWIBMNLQWZKDTISKM3ANCNFSM4WKD5GOA>
> .
>


-- 
Regards,
Ashish Bhat

sagetherage commented:
@pyashish if we need to request a repo to host it within this project org, we can, it may take a bit of time to get approval, but I can start that process. do you wish to use the Salt Test Suite as the module is, over time, developed?

pyashish commented:
@sagetherage, This is just a sample test suite we have written so that if there are no more comments on implementation we can replicate for the rest. 

The goal is to natively get listed in SaltStack and become a native package , we are ok to host the repo in our Org. 
Also we want one-cmd to install our modules , something like pip install cohesity_management_sdk  or ansible-galaxy install cohesity .

naveena-maplelabs commented:
@s0undt3ch @sagetherage 
Can you please review the sample testcase added.

sagetherage commented:
@naveena-maplelabs @pyashish the test case needs some help, but if you are going to put this into a Salt-Extension in your own org then it does NOT need to be merged, here. You are welcome to close this and merge it into your own GitHub or GitLab (or other) organization and repository. 

We have a tool in another repo here: https://github.com/saltstack/salt-extension to make an extension and you can also see the [overview video](https://youtu.be/hhomJkwxK3Q).

If you have questions please reach out to us on [Slack](https://saltstackcommunity.slack.com) or IRC via [Freenode](https://webchat.freenode.net/#salt) or you can join our mailing list/forum for [salt-users](https://groups.google.com/g/salt-users).

naveena-maplelabs commented:
Thank you for the suggestions and necessary docs. @sagetherage 
We are planning to go with salt extension.

naveena-maplelabs commented:
@sagetherage I have installed salt-extension package and created a module extension for our module. How can i use the module extended, its not getting listed when i run "salt --versions" command.

s0undt3ch commented:
I you created an extension using https://github.com/saltstack/salt-extension, after adding your module(s) to your newly created extensions, you will have to install it on the same python environment as salt, and restart the daemons(master, minion), at which time, the extension should become visible to salt.

naveena-maplelabs commented:
@s0undt3ch, After adding modules to ~/src/salttext/cohesity/modules generated using create-salt-extension, how to install the module extension.

s0undt3ch commented:
Your new extension is a regular python package, `python -m pip install ~/src/salttext/cohesity/modules`

sagetherage commented:
Closing this as we will not be merging it into the Salt Code base.

naveena-maplelabs commented:
@sagetherage @s0undt3ch  we can created an salt extension with our modules added. But is there any way we can get listed similar to this, https://docs.saltproject.io/en/latest/ref/modules/all/index.html

s0undt3ch commented:
Make sure you populate your docs, like salt does, with stubs, and then run `nox -e docs-html`
That should do it.

naveena-maplelabs commented:
thanks @s0undt3ch. I was able to generate the html files. Is there any way to get listed here https://docs.saltproject.io/en/latest/ref/modules/all/index.html
